File: osm/addr-osmify-java/src/main/java/hu/vmiklos/addr_osmify/DefaultUrlopener.java
/*
 * Copyright 2019 Miklos Vajna
 *
 * SPDX-License-Identifier: MIT
 */

package hu.vmiklos.addr_osmify;

import java.io.OutputStream;
import java.io.StringReader;
import java.io.StringWriter;
import java.net.HttpURLConnection;
import java.net.URL;
import org.apache.commons.io.IOUtils;

/**
 * Default urlopen(), using HttpURLConnection.
 */
public class DefaultUrlopener implements Urlopener
{
    public String urlopen(String urlString, String data) throws Exception
    {
        URL url = new URL(urlString);
        HttpURLConnection connection = (HttpURLConnection)url.openConnection();

        if (data.isEmpty())
        {
            StringWriter writer = new StringWriter();
            IOUtils.copy(connection.getInputStream(), writer);
            return writer.toString();
        }

        StringReader reader = new StringReader(data);
        OutputStream outputStream = null;
        try
        {
            connection.setRequestMethod("POST");
            connection.setDoOutput(true);
            outputStream = connection.getOutputStream();
            IOUtils.copy(reader, outputStream);

            StringWriter writer = new StringWriter();
            IOUtils.copy(connection.getInputStream(), writer);
            return writer.toString();
        }
        catch (Exception e)
        {
            throw e;
        }
        finally
        {
            reader.close();
            if (outputStream != null)
            {
                outputStream.close();
            }
        }
    }
}

/* vim:set shiftwidth=4 softtabstop=4 expandtab: */

File: osm/addr-osmify-kotlin/src/main/kotlin/DefaultUrlopener.kt
/*
 * Copyright 2020 Miklos Vajna. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */
package hu.vmiklos.addr_osmify

import org.apache.commons.io.IOUtils
import java.io.OutputStream
import java.io.StringReader
import java.io.StringWriter
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.Charset

/**
 * Default urlopen(), using HttpURLConnection.
 */
class DefaultUrlopener : Urlopener {
    override fun urlopen(urlString: String, data: String): String {
        val url = URL(urlString)
        val connection = url.openConnection() as HttpURLConnection
        val charset: Charset? = null
        if (data.isEmpty()) {
            val writer = StringWriter()
            IOUtils.copy(connection.inputStream, writer, charset)
            return writer.toString()
        }
        val reader = StringReader(data)
        var outputStream: OutputStream? = null
        return try {
            connection.requestMethod = "POST"
            connection.doOutput = true
            outputStream = connection.outputStream
            IOUtils.copy(reader, outputStream, charset)
            val writer = StringWriter()
            IOUtils.copy(connection.inputStream, writer, charset)
            writer.toString()
        } catch (e: Exception) {
            throw e
        } finally {
            reader.close()
            outputStream?.close()
        }
    }
}

/* vim:set shiftwidth=4 softtabstop=4 expandtab: */
